docs/source/models/spatial-knn.rst

```diff
@@ -0,0 +1,288 @@
+============================
+
+Intro
+###################
+
+Nearest neighbour problem is defined as finding an observation from a set S that is the most similar to the given
+observation O. The similarity is defined by a distance function d(O, S[i]). The most common distance function is the Euclidean.
+The nearest neighbour problem is a special case of the k-nearest neighbour problem, where k=1. The k-nearest neighbour
+problem is defined as finding k observations from a set S that are the most similar to the given observation O.
+
+.. figure:: ../images/nearest_neighbour.png
+   :figclass: doc-figure
+
+   Fig 1. Nearest neighbour problem
+
+We define spatial k-nearest neighbour problem as finding k observations from a set S that are the most similar to the
+given observation O, where the similarity is defined by a distance function d(O, S[i]) is defined by st_distance(O, S[i]) function.
+In our case we further define the similarity as the inverse of the distance, so the most similar observation is the one with
+the smallest distance. ST_Distance function is defined as the shortest distance between two geometries in projected units.
+We do not restrict the type of geometries that can be used in the problem. The only requirement is that the geometries
+must be in the same coordinate system.
+
+Spatial KNN as a Join Relation
+##############################
+
+The traditional definition of the nearest neighbour problem is defined as:
+"given a set S of points in a space M and a query point q in M, find the point in S that is closest to q".
+We have relaxed this definition by allowing the query point to be a set of points. The result of the nearest neighbour
+problem is a set of points from S that are closest to each query points. This in effect formulates a join condition
+where a point pair (Q[i], S[j]) is a match if st_distance(Q[i], S[j]) <= st_distance(Q[i], S[k]) where S[k] is the
+kth nearest neighbour of Q[i].
+
+The problem of finding the k-nearest neighbours can be formulated as a join condition where a point pair (Q[i], S[j]) is
+the fact that this is not an equality relation and the join condition is not symmetric. If S[j] is in the knn set of Q[i]
+it does not imply that Q[i] is in the knn set of S[j]. This join would default to a Cartesian product if we were to
+use the traditional join syntax. We have to use a different approach.
+
+Spatial KNN as an Iterative Transformation
+##########################################
+
+The problem of finding the k-nearest neighbours can be formulated as an iterative transformation. The transformation
+is defined as follows:
+* For each geometry in set Q generate a hex ring in grid index space
+* Generate match candidates in the hex ring
+* For each match candidate calculate the distance to the query geometry
+* For each geometry in set Q count the matches and stop if the count is equal to k
+* Continue with the next hex ring until max number of iterations is performed
```

Comment:
Do you always continue until max number of iterations, or do you stop if all Q have at least k neighbours?